src/Bicep.Cli/Services/TemplateEvaluatorService.cs

```diff
@@ -0,0 +1,239 @@
+                    {
+                        return config.OnReferenceFunc(resourceId, apiVersion, fullBody);
+                    }
+
+                    if (resourceLookup.TryGetValue(resourceId, out var foundResource) &&
+                        (apiVersion is null || StringComparer.OrdinalIgnoreCase.Equals(apiVersion, foundResource.ApiVersion.Value)))
+                    {
+                        return fullBody ? foundResource.ToJToken() : foundResource.Properties.ToJToken();
+                    }
+                }
+
+                var value = defaultEvaluateFunction(functionExpression, parameters, additionalInfo);
+                return value;
+            };
+
+            for (int i = 0; i < template.Resources.Length; i++)
+            {
+                var resource = template.Resources[i];
+
+                if (resource.Properties is not null)
+                {
+                    var skipEvaluationPaths = new InsensitiveHashSet();
+                    if (resource.Type.Value.EqualsOrdinalInsensitively("Microsoft.Resources/deployments"))
+                    {
+                        skipEvaluationPaths.Add("template");
+                    };
+
+                    resource.Properties.Value = ExpressionsEngine.EvaluateLanguageExpressionsRecursive(
+                        root: resource.Properties.Value,
+                        evaluationContext: evaluationContext,
+                        skipEvaluationPaths: skipEvaluationPaths);
+                }
+            }
+
+            if (template.Outputs is not null && template.Outputs.Count > 0)
+            {
+                foreach (var outputKey in template.Outputs.Keys.ToList())
+                {
+                    template.Outputs[outputKey].Value.Value = ExpressionsEngine.EvaluateLanguageExpressionsRecursive(
+                        root: template.Outputs[outputKey].Value.Value,
+                        evaluationContext: evaluationContext);
+                }
+            }
+        }
+
+        public static JToken Evaluate(JToken? templateJtoken, JToken? parametersJToken = null, Func<EvaluationConfiguration, EvaluationConfiguration>? configBuilder = null)
+        {
+            var configuration = EvaluationConfiguration.Default;
+
+            if (configBuilder is not null)
+            {
+                configuration = configBuilder(configuration);
+            }
+
+            return EvaluateTemplate(templateJtoken, parametersJToken, configuration);
+        }
+
+        private static JToken EvaluateTemplate(JToken? templateJtoken, JToken? parametersJToken, EvaluationConfiguration config)
+        {
+            templateJtoken = templateJtoken ?? throw new ArgumentNullException(nameof(templateJtoken));
+
+            var deploymentScope = GetDeploymentScope(templateJtoken["$schema"]!.ToString());
+
+            var metadata = new InsensitiveDictionary<JToken>(config.Metadata);
```

Comment:
Should I keep the metadata this way or should I send an empty metadata. 

I believe in this example I am actually sending the metadata as an empty dictionary